crates/biome_css_analyze/src/lint/nursery/no_descending_specificity.rs

```diff
@@ -0,0 +1,196 @@
+    /// a { color: red; }
+    /// ```
+    ///
+    /// ```css,expect_diagnostic
+    /// a {
+    ///   & > b { color: red; }
+    /// }
+    /// b { color: red; }
+    /// ```
+    ///
+    /// ```css,expect_diagnostic
+    /// :root input {
+    ///     color: red;
+    /// }
+    /// html input {
+    ///     color: red;
+    /// }
+    /// ```
+    ///
+    ///
+    /// ### Valid
+    ///
+    /// ```css
+    /// a { color: red; }
+    /// b a { color: red; }
+    /// ```
+    ///
+    /// ```css
+    /// b { color: red; }
+    /// a {
+    ///   & > b { color: red; }
+    /// }
+    /// ```
+    ///
+    /// ```css
+    /// a:hover { color: red; }
+    /// a { color: red; }
+    /// ```
+    ///
+    pub NoDescendingSpecificity {
+        version: "next",
+        name: "noDescendingSpecificity",
+        language: "css",
+        recommended: false,
+        sources: &[RuleSource::Stylelint("no-descending-specificity")],
+    }
+}
+
+#[derive(Debug)]
+pub struct DescendingSelector {
+    high: (TextRange, Specificity),
+    low: (TextRange, Specificity),
+}
+/// find tail selector
+/// /// ```css
+/// a b:hover {
+///   ^^^^^^^
+/// }
+/// ```
+fn find_tail_selector(selector: &AnyCssSelector) -> Option<String> {
+    match selector {
+        AnyCssSelector::CssCompoundSelector(s) => {
+            let simple = s
+                .simple_selector()
+                .map_or(String::new(), |s| s.syntax().text_trimmed().to_string());
+            let sub = s.sub_selectors().syntax().text_trimmed().to_string();
+
+            let last_selector = [simple, sub].join("");
+            Some(last_selector)
+        }
+        AnyCssSelector::CssComplexSelector(s) => {
+            s.right().as_ref().ok().and_then(find_tail_selector)
+        }
+        _ => None,
+    }
+}
+
+fn find_descending_selector(
+    rule: &CssSemanticRule,
+    model: &SemanticModel,
+    visited_rules: &mut BTreeSet<RuleId>,
+    visited_selectors: &mut BTreeMap<String, (TextRange, Specificity)>,
+    descending_selectors: &mut Vec<DescendingSelector>,
+) {
+    if visited_rules.contains(&rule.id) {
+        return;
+    } else {
+        visited_rules.insert(rule.id);
+    };
+
+    for selector in &rule.selectors {
+        let tail_selector = if let Some(s) = find_tail_selector(&selector.original) {
+            s
+        } else {
+            continue;
+        };
+
+        if let Some((last_textrange, last_specificity)) = visited_selectors.get(&tail_selector) {
+            if last_specificity > &selector.specificity {
+                descending_selectors.push(DescendingSelector {
+                    high: (*last_textrange, last_specificity.clone()),
+                    low: (selector.range, selector.specificity.clone()),
+                });
+            }
+        } else {
+            visited_selectors.insert(
+                tail_selector,
+                (selector.range, selector.specificity.clone()),
+            );
+        }
```

Comment:
consider adding some comments here.

Reply:
@chansuke Thank you for the feedback. Usually, you want to provide an actionable comment, maybe with some advice. In this case, what kind of comments should OP add? Is it because the logic you highlighted is hard read? Is it because it's too complex? Let's help our contributors ;)

Reply:
https://github.com/biomejs/biome/blob/c3b8b5be6ef0253e1e982d2f3a925a39099cdc56/crates/biome_css_analyze/src/lint/nursery/no_descending_specificity.rs#L111-L116
Added comment. How about like this?